### PR TITLE
[ark-multiplayer] feat: Set up basic multiplayer networking

### PR DESCRIPTION
**Pull request by ryanpeh**
## Features

Implement multiplayer communication. Messages are broadcasted between devices through P2P.

Currently only supports sending of events to peers. We need to find some way to implement state syncing with consistency.

Event sending requires that the events conform to `ArkSerializableEvent`, and only those events will be broadcasted to other users.

## Changes

Changed implementation of `ArkEventManager`
- Subscribing is now through `Event.Type` rather than `Event.id`

### CODE REVIEW

File: ArkKit/ark-event-kit/ArkEvent.swift [hunk 1]
```diff
@@ -7,12 +7,18 @@
...
 
-typealias ArkEventID = UUID
+typealias ArkEventID = String
 
-protocol ArkEvent<Data> {
+protocol ArkEvent<Data>: Codable {
     associatedtype Data = ArkEventData
 
-    static var id: ArkEventID { get }
+    static var id: String { get }
```
Comment (markusyeo) [inline ArkKit/ark-event-kit/ArkEvent.swift:15]: Should this be ArkEventID? if not line 10 is redundant

File: ArkKit/ark-event-kit/DemoArkEvent.swift [hunk 2]
```diff
@@ -5,49 +5,49 @@
...
-            guard let eventData = event.eventData as? DemoArkEventData else {
-                return
-            }
-            // Print event data for demonstration purposes.
-            print(eventData.name)
-            print(eventData.number)
-        }
-    }
-}
+// import Foundation
```
Comment (markusyeo) [inline ArkKit/ark-event-kit/DemoArkEvent.swift:8]: To be updated

File: ArkKit/Ark.swift [hunk 3]
```diff
@@ -49,12 +49,17 @@ class Ark {
         self.blueprint = blueprint
         let eventManager = ArkEventManager()
         let ecsManager = ArkECS()
+//        let multiplayerManager = ArkMultiplayerManager(serviceName: "ArkMultiplayer")
```
Comment (markusyeo) [inline ArkKit/Ark.swift:52]: Do we still need these?

File: ArkKit/ark-multiplayer-kit/ArkMultiplayerEventManager.swift [hunk 4]
```diff
@@ -0,0 +1,40 @@
+//
+//  ArkMultiplayerEventManager.swift
+//  ArkKit
+//
+//  Created by Ryan Peh on 31/3/24.
+//
+
+import Foundation
+
+class ArkMultiplayerEventManager: ArkEventManager {
```
Comment (markusyeo) [inline ArkKit/ark-multiplayer-kit/ArkMultiplayerEventManager.swift:10]: Do we still need this class too?

File: ArkKit/ark-event-kit/ArkEvent.swift (hunk 1 above)
Comment (ryanpeh) [reply ArkKit/ark-event-kit/ArkEvent.swift]: Removed `id`

File: ArkKit/Ark.swift (hunk 3 above)
Comment (ryanpeh) [reply ArkKit/Ark.swift]: Reworked multiplayer integration

File: ArkKit/ark-event-kit/DemoArkEvent.swift (hunk 2 above)
Comment (ryanpeh) [reply ArkKit/ark-event-kit/DemoArkEvent.swift]: Removed file

File: ArkKit/ark-event-kit/ark-event-data/ArkEventData.swift [hunk 5]
```diff
@@ -5,6 +5,6 @@
 //  Created by Ryan Peh on 13/3/24.
 //
 
-protocol ArkEventData {
+protocol ArkEventData: Codable {
```
[the remark this answers is not in the document]
Comment (ryanpeh) [reply ArkKit/ark-event-kit/ark-event-data/ArkEventData.swift]: Removed Codable requirement from ArkEvent / ArkEventData, implemented new `ArkSerializableEvent`

**Comment by didymental** [comment]
for consideration on syncing between diff peers:
- use one peer as the centralised authority of sorts to sync to? can be the "host" of the game.

**Review by didymental — APPROVED** [review]
lgtm, to implement state sync in future PR

**Pull request merged**